File: rust/src/day02.rs
use crate::util;

struct Command {
    operation: String,
    arg: i64
}

pub fn solve() {
    let commands = get_commands();
    let p = navigate_simple(&commands);

    println!("=====  Day 2  =====");
    println!("The final position is {:?} with the product {}", p, p.0 * p.1);
}

fn navigate_simple(commands: &[Command]) -> (i64, i64) {
    let mut x = 0;
    let mut y = 0;
    for cmd in commands {
         match cmd.operation.as_str() {
            "forward" => x += cmd.arg,
            "down" => y += cmd.arg,
            "up" => y -= cmd.arg,
            anything => panic!("Unknown command {}", anything)
         }
    }
    return (x, y);
}


fn get_commands() -> Vec<Command> {
    let data = util::read_data("day02");
    let mut commands = Vec::new();

    let lines: Vec<&str> = data.split("\n").collect();
    for line in lines {
        let parts: Vec<&str> = line.split(" ").collect();
        let cmd = Command { operation: parts[0].to_string(), arg: parts[1].parse().unwrap() };
        commands.push(cmd);
    }

    return commands;
}

File: typescript/src/days/day02.ts
import { readFileSync } from "fs";

type Command = {
  cmd: string;
  arg: number;
};

export function day02() {
    const cmds = getData();
    const p1 = run(cmds);
    const p2 = runWithAim(cmds);
    
    console.log(`The submarine ends up at ${p1.x},${p1.y} with the product`, p1.x * p1.y);
    console.log(`With aiming it ends up at ${p2.x},${p2.y} with the product`, p2.x * p2.y);
}

function run(cmds: Command[]) {
    let x = 0;
    let y = 0;
    for (let {cmd, arg} of cmds) {
        switch(cmd) {
            case 'forward': x += arg; break;
            case 'down': y += arg; break;
            case 'up': y -= arg; break;
            default: console.log('Unknown command', cmd);
        }
    }
    return {x, y};
}

function runWithAim(cmds: Command[]) {
    let x = 0;
    let y = 0;
    let aim = 0;
    for (let {cmd, arg} of cmds) {
        switch(cmd) {
            case 'forward': x = x + arg; y += aim * arg; break;
            case 'down': aim += arg; break;
            case 'up': aim -= arg; break;
            default: console.log('Unknown command', cmd);
        }
    }
    return {x, y};
}

function getData(): Command[] {
  return readFileSync("../data/day02.txt", "utf-8")
    .split("\n")
    .map((line) => {
      const [cmd, arg] = line.split(" ");
      return { cmd, arg: +arg };
    });
}


